File: examples/go-vuln-demo/src/p2_counter_evidence.go
package main

import (
	"sync"
	"sync/atomic"
)

type ResourceHandle struct {
	buffer []byte
}

// P2-01: defer Close — 非泄漏
func safeResource() {
	h := &ResourceHandle{buffer: make([]byte, 1024)}
	defer func() { h.buffer = nil }()
	_ = len(h.buffer)
}

var counter int
var counterMu sync.Mutex

// P2-02: defer Unlock — 非竞争
func safeIncrement() int {
	counterMu.Lock()
	defer counterMu.Unlock()
	counter++
	return counter
}

// P2-03: atomic 操作
var atomicCounter int64

func safeAtomic() int64 {
	return atomic.AddInt64(&atomicCounter, 1)
}

type SafeWriter struct {
	mu sync.Mutex
}

// P2-04: 结构体封装锁
func (w *SafeWriter) Write(data string) {
	w.mu.Lock()
	defer w.mu.Unlock()
}

File: examples/java-vuln-demo-no-answers/src/com/secguardian/demo/p2_counter_evidence.java
package com.secguardian.demo;


import java.util.concurrent.locks.*;
import java.util.concurrent.atomic.*;

class ResourceHandle implements AutoCloseable {
    private byte[] buffer = new byte[1024];
    public void close() { buffer = null; }  
}

public class p2_counter_evidence {
    private final Lock mutex = new ReentrantLock();
    private final AtomicInteger counter = new AtomicInteger(0);
    private int sharedState = 0;
    private int fileHandle = 0;


    public void safeResource() {
        try (ResourceHandle h = new ResourceHandle()) {
            
        }
    }


    public void safeIncrement() {
        mutex.lock();
        try {
            sharedState++;
        } finally {
            mutex.unlock();
        }
    }


    public int safeCounter() {
        return counter.incrementAndGet();
    }


    public synchronized void syncWrite(String data) {
        
    }
}
